File: src/ipc/messages/ValueBase.ts
import { ValueType } from './valuetype';
import { LeapError } from './LeapError';
import * as ldm from './leapdatamodel';

export abstract class ValueBase {
    abstract _verifyAsTypeOrRaise(path: string[], valueType: ValueType): void;
    abstract _toLdm(valueType: ValueType): ldm.Value;

    static _verifyOrRaise(value: any, valueType: ValueType): void {
        ValueBase._verifyValueOrRaise([], value, valueType);
    }

    static _verifyValueOrRaise(path: string[], value: any, valueType: ValueType): void {
        if (valueType.name == 'str') {
            if (!((typeof value === 'string') || (value instanceof String))) {
                throw ValueBase._pathError(path, 'expecting string');
            }
        } else if (valueType.name === 'int') {
            if (!((typeof value === 'number') || (value instanceof Number))) {
                throw ValueBase._pathError(path, 'expecting integer');
            }
        } else if (valueType.name === 'float') {
            if (!((typeof value === 'number') || (value instanceof Number))) {
                throw ValueBase._pathError(path, 'expecting float');
            }
        } else if (valueType.name === 'bool') {
            if (!((typeof value === 'boolean') || (value instanceof Boolean))) {
                throw ValueBase._pathError(path, 'expecting boolean');
            }
        } else if (valueType.name === 'list') {
            if (value instanceof Array) {
                ValueBase._verifyListOrRaise(path, value, valueType.args[0]);
            } else {
                throw ValueBase._pathError(path, 'expecting list');
            }
        } else if (valueType.name === 'option' && valueType.args[0].name !== 'option') {
            if (value !== null) {
                ValueBase._verifyValueOrRaise(path, value, valueType.args[0]);
            }
        } else if (value instanceof ValueBase) {
            value._verifyAsTypeOrRaise(path, valueType);
        }
        else {
            throw ValueBase._pathError(path, `expecting ${valueType}`);
        }
    }

    protected static _pathError(path: string[], message: string): LeapError {
        if (path.length === 0) {
            return new LeapError(message);
        } else {
            const p = path.join('.');
            return new LeapError(`${p}: ${message}`);
        }
    }

    static _verifyListOrRaise(path: string[], values: any[], elementType: ValueType): void {
        values.forEach((v, i) => ValueBase._verifyValueOrRaise(path.concat([i.toString()]), v, elementType));
    }

    static _valueToLdm(value: any, valueType: ValueType): ldm.Value {
        if (valueType.name === 'str') {
            return new ldm.StringValue(String(value));
        } else if (valueType.name === 'int') {
            return new ldm.IntegerValue(Math.trunc(value));
        } else if (valueType.name === 'float') {
            return new ldm.FloatValue(Number(value));
        } else if (valueType.name === 'bool') {
            return new ldm.BooleanValue(Boolean(value));
        } else if (valueType.name === 'list') {
            return new ldm.ListValue(
                (value as Array<any>).map(v => ValueBase._valueToLdm(v, valueType.args[0])),
                valueType
            );
        } else if (valueType.name === 'option' && valueType.args[0].name !== 'option') {
            if (value === null) {
                return new ldm.EnumValue(
                    'none',
                    new ldm.StructValue(new Map(), new ValueType('none', [])),
                    valueType
                );
            } else {
                let valuesMap = new Map();
                valuesMap.set('value', ValueBase._valueToLdm(value, valueType.args[0]));
                return new ldm.EnumValue(
                    'some',
                    new ldm.StructValue(valuesMap, valueType.args[0]),
                    valueType
                );
            }
        } else if (value instanceof ValueBase) {
            return value._toLdm(valueType);
        }
        throw new LeapError('unexpected value type');
    }
};

File: src-tauri/src/ipc/messages/value_base.rs
use super::leap_data_model as ldm;
use super::value_type::ValueType;
use std::collections::HashMap;

pub trait ValueBase {
    const _TYPE: &'static str;
    fn _type(&self) -> &str {
        Self::_TYPE
    }
    fn _applied_args() -> Vec<ValueType>;
    fn _value_type(&self) -> ValueType {
        ValueType::new(Self::_TYPE.to_owned(), Self::_applied_args())
    }
    fn _to_ldm(&self) -> ldm::Value;
}

impl ValueBase for String {
    const _TYPE: &'static str = "str";

    fn _applied_args() -> Vec<ValueType> {
        vec![]
    }

    fn _to_ldm(&self) -> ldm::Value {
        ldm::Value::String(ldm::StringValue(self.clone()))
    }
}

impl ValueBase for i64 {
    const _TYPE: &'static str = "int";

    fn _applied_args() -> Vec<ValueType> {
        vec![]
    }

    fn _to_ldm(&self) -> ldm::Value {
        ldm::Value::Integer(ldm::IntegerValue(*self))
    }
}

impl ValueBase for f64 {
    const _TYPE: &'static str = "float";

    fn _applied_args() -> Vec<ValueType> {
        vec![]
    }

    fn _to_ldm(&self) -> ldm::Value {
        ldm::Value::Float(ldm::FloatValue(*self))
    }
}

impl ValueBase for bool {
    const _TYPE: &'static str = "bool";

    fn _applied_args() -> Vec<ValueType> {
        vec![]
    }

    fn _to_ldm(&self) -> ldm::Value {
        ldm::Value::Boolean(ldm::BooleanValue(*self))
    }
}

impl<T: ValueBase> ValueBase for Vec<T> {
    const _TYPE: &'static str = "list";

    fn _applied_args() -> Vec<ValueType> {
        vec![ValueType::new(T::_TYPE.to_owned(), T::_applied_args())]
    }

    fn _to_ldm(&self) -> ldm::Value {
        let value = self.iter().map(|v| v._to_ldm()).collect();
        ldm::Value::List(ldm::ListValue {
            value,
            value_type: ValueType::new(Self::_TYPE.to_owned(), Self::_applied_args()),
        })
    }
}

impl<T: ValueBase> ValueBase for Option<T> {
    const _TYPE: &'static str = "option";

    fn _applied_args() -> Vec<ValueType> {
        vec![ValueType::new(T::_TYPE.to_owned(), T::_applied_args())]
    }

    fn _to_ldm(&self) -> ldm::Value {
        if let Some(v) = self {
            let mut value = HashMap::new();
            value.insert("value".to_owned(), v._to_ldm());
            ldm::Value::Enum(ldm::EnumValue {
                variant: "some".to_owned(),
                value: ldm::StructValue {
                    value,
                    value_type: ValueType::new("some".to_owned(), Self::_applied_args()),
                },
                value_type: ValueType::new("option".to_owned(), Self::_applied_args()),
            })
        } else {
            ldm::Value::Enum(ldm::EnumValue {
                variant: "none".to_owned(),
                value: ldm::StructValue {
                    value: HashMap::new(),
                    value_type: ValueType::new("none".to_owned(), vec![]),
                },
                value_type: ValueType::new("option".to_owned(), Self::_applied_args()),
            })
        }
    }
}

impl<T: ValueBase, E: ValueBase> ValueBase for Result<T, E> {
    const _TYPE: &'static str = "result";

    fn _applied_args() -> Vec<ValueType> {
        vec![
            ValueType::new(T::_TYPE.to_owned(), T::_applied_args()),
            ValueType::new(E::_TYPE.to_owned(), E::_applied_args()),
        ]
    }

    fn _to_ldm(&self) -> ldm::Value {
        let mut value = HashMap::new();
        match self {
            Ok(v) => {
                value.insert("value".to_owned(), v._to_ldm());
                ldm::Value::Enum(ldm::EnumValue {
                    variant: "ok".to_owned(),
                    value: ldm::StructValue {
                        value,
                        value_type: ValueType::new(
                            "ok".to_owned(),
                            vec![ValueType::new(T::_TYPE.to_owned(), T::_applied_args())],
                        ),
                    },
                    value_type: ValueType::new("result".to_owned(), Self::_applied_args()),
                })
            }
            Err(e) => {
                value.insert("value".to_owned(), e._to_ldm());
                ldm::Value::Enum(ldm::EnumValue {
                    variant: "err".to_owned(),
                    value: ldm::StructValue {
                        value,
                        value_type: ValueType::new(
                            "err".to_owned(),
                            vec![ValueType::new(E::_TYPE.to_owned(), E::_applied_args())],
                        ),
                    },
                    value_type: ValueType::new("result".to_owned(), Self::_applied_args()),
                })
            }
        }
    }
}

impl<T: ValueBase> ValueBase for Box<T> {
    const _TYPE: &'static str = T::_TYPE;

    fn _applied_args() -> Vec<ValueType> {
        T::_applied_args()
    }

    fn _to_ldm(&self) -> ldm::Value {
        T::_to_ldm(self)
    }
}
